src/js/utils/linked-list.js

```diff
@@ -0,0 +1,84 @@
+export default class LinkedList {
+  constructor(options) {
+    this.head = null;
+    this.tail = null;
+    if (options) {
+      let {adoptItem, freeItem} = options;
+      this.adoptItem = adoptItem;
+      this.freeItem = freeItem;
+    }
+  }
+  prepend(item) {
+    this.insertBefore(item, this.head);
+  }
+  append(item) {
+    this.insertBefore(item, null);
+  }
+  insertAfter(item, prevItem) {
+    let nextItem = null;
+    if (prevItem) {
+      nextItem = prevItem.next;
+    }
+    this.insertBefore(item, nextItem);
+  }
+  insertBefore(item, nextItem) {
```

Comment:
I find these arg names mildly confusing — `insertBefore(newItem, oldItem)` makes more sense to me.

I think this can be simplified:
```
newItem.prev = oldItem.prev;
newItem.next = oldItem;

if (oldItem.prev) {
  oldItem.prev.next = newItem;
} else {
  this.head = newItem;
}

oldItem.prev = newItem;
```